aocd/models.py

```diff
@@ -278,48 +327,70 @@ def title(self):
         """Hook for IPython's pretty-printer."""
-        if cycle:
-            p.text(repr(self))
-        else:
-            txt = f"<Puzzle({self.year}, {self.day}) at {hex(id(self))} - {self.title}>"
-            p.text(txt)
+        txt = repr(self) if cycle else f"<Puzzle({self.year}, {self.day}) at {hex(id(self))} - {self.title}>"
+        p.text(txt) # type: ignore[no-untyped-call] # IPython/Jupyter doesn't provide type hints
 
-    def _coerce_val(self, val):
+    def _coerce_val(self, val: _Answer) -> str:
         # technically adventofcode.com will only accept strings as answers.
         # but it's convenient to be able to submit numbers, since many of the answers
         # are numeric strings. coerce the values to string safely.
-        orig_val = val
-        orig_type = type(val)
+
+        original_val = val
+        def fail() -> NoReturn:
+            msg = f"Failed to coerce {type(original_val).__name__} value {original_val!r} for {self.year}/{self.day:02}."
+            raise AocdError(msg)
+
         coerced = False
-        floatish = isinstance(val, (float, complex))
-        if floatish and val.imag == 0.0 and val.real.is_integer():
-            coerced = True
-            val = int(val.real)
-        elif orig_type.__module__ == "numpy" and getattr(val, "ndim", None) == 0:
-            # deal with numpy scalars
-            if orig_type.__name__.startswith(("int", "uint", "long", "ulong")):
+
+        # if we get an ImportError, numpy is not installed, so we skip handling numpy types
+        with contextlib.suppress(ImportError):
+            import numpy as np
+
+            # "unwrap" arrays that contain a single element
+            if isinstance(val, np.ndarray) and val.size == 1:
                 coerced = True
-                val = int(orig_val)
-            elif orig_type.__name__.startswith(("float", "complex")):
-                if val.imag == 0.0 and float(val.real).is_integer():
-                    coerced = True
-                    val = int(val.real)
+                val = val.item()
+            if isinstance(val, (np.integer, np.floating, np.complexfloating)) and val.imag == 0 and val.real.is_integer():
+                coerced = True
+                val = str(int(val.real))
         if isinstance(val, int):
             val = str(val)
+        elif isinstance(val, (float, complex)) and val.imag == 0 and val.real.is_integer():
+            coerced = True
+            val = str(int(val.real))
+        elif isinstance(val, bytes):
+            coerced = True
+            val = val.decode()
+        elif isinstance(val, (Decimal, Fraction)):
+            # if val can be represented as an integer ratio where the denominator is 1
+            # val is an integer and val == numerator
+            numerator, denominator = val.as_integer_ratio()
+            if denominator == 1:
+                coerced = True
+                val = str(numerator)
```

Comment:
Although it's a new feature (or arguably a bugfix), this is somewhat related to the typing changes because the goal was for `_coerce_val` to be annotated with `str` return type. Without the refactor done here, the return type could be pretty much anything..

Reply:
I.. don't quite see that. If the type hints state that only specific number types are accepted and are handled then any other types passed in (and not resulting in an exception or a str return value) are simply out of scope.